<h1 style="color:rgb(0,120,191);">Bayesian Classifiers</h1>
<br/>
<p>
    In this tutorial we will explore a machine learning algorithm used for classification problems: <strong style="color:rgb(0,120,191);">the Bayes classifier</strong> and its variants, <strong style="color:rgb(0,120,191);">the Plug-In and Naïve Bayes classifiers</strong>.
    We will give proper definition of each of these and apply the algorithms on a toy dataset, and also on the well-known iris dataset.
</p>
<p> 
    <strong>Prerequisites</strong>:<br/> knowledge of following machine learning concepts: feature, label, training data, validation data, accuracy, maximum likelihood estimation 
</p>


<h1 style="color:rgb(0,120,191);">1) What is classification?</h1>
<h2>a) The classification problem</h2>
<p>
    We have n observations x<sub>1</sub>, ..., x<sub>n</sub> in R<sup>d</sup> and we would like to assign each of these vectors into a class  (i.e a category) among K classes y<sub>1</sub>, y<sub>2</sub>, ..., y<sub>k</sub>. <br/>
If K=2, we talk about <strong>binary classification</strong>, for K &ge; 3 we talk about <strong>multiclass classification.</strong>
</p>

<p>
    Let f be the function which maps input vector x to a class y:  &nbsp;    <strong>y = f(x)</strong> <br/>
    <ol>
        We need to
        <li>
             Define what function we're going to use for this mapping </li>
        <li>Learn the parameters of this function by using pairs of labeled points (x<sub>i</sub>,y<sub>i</sub>) i=1,...,n</li>
        </ol>
    
</p>

<h2>b) How can we measure the quality of a classifier?</h2>

<ul>
    <li>
        The prediction accuracy is the probability of classifying correctly the data: P( f(x) = y ). We want to maximize this!       </li>
    <li>The prediction error is the probability of misclassifying the data: P( f(x) &#x2260; y ). We want to minimize this!</li>
</ul>

<p>
    To calculate these probabilities, we need to make an important hypothesis here: <br/>
    
   Our observed data (x<sub>i</sub>,y<sub>i</sub>)  i=1,..., n are <strong>independant and identically distributed (IID)</strong> via a distribution D (that we dont know).<br/>
   "Independent" means that the observation x<sub>i</sub> does not depend on the observation x<sub>j</sub> for any i &ne; j. <br/>
   "Identically distributed" means that the samples come from the same distribution. <br/><br/>
   <strong>NB: Through the statistical learning theory</strong>, we know that <strong>performing well on the training set ensures a good generalization on the testing set!</strong> (where we only get to see the x<sub>i</sub>)
</p>

<br/><h1 style="color:rgb(0,120,191);">2) What is the Bayes classifier?</h1>
<p>
    Let's define the optimal classifier as the classifier which minimizes the prediction error. <br/>
    If we get the data IID from some underlying distribution D, and then for every single vector x, we predict the label y to be the most probable label according to that distribution, then the prediction error will be minimized and <strong style="color:rgb(0,120,191);">this classifier is the Bayes classifier!</strong>
</p>
<p>
    But of course, we don't know the distribution D, so we will have to approximate it, and <strong>therefore we will no longer be able to say the classifier is optimal!</strong>
</p>
<p>
  So what we want, is to maximize the probability P(y|x), &nbsp;&nbsp;
  Let's recall Bayes rule: <br/> &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;
  P(y|x) = P(x|y) P(y)  /  P(x) &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;(E1) 
  <ul>
    <li>P(y|x) is called the <strong>posterior</strong>, it is the posterior belief of the unknown class y given the observation x</li>
    <li>P(x|y) is the <strong>class conditionnal distribution</strong> (also called the <strong>likelihood</strong> of the data given the class), it is the class specific distribution on vector x given that it belongs to class y</li>
    <li>P(y) is the class <strong>prior</strong>, it's the a priori assumption we make on the distribution of the class</li> 
    <li>P(x) is called the <strong>evidence</strong>, it is a normalizing constant</li>      
  </ul>
So we want to maximize the probability of the posterior:   we're interested in finding the class y which is the most probable, so we can write this: <br/>    
&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; argmax<sub>y</sub> P(y|x) &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; <=>  &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; argmax<sub>y</sub> P(x|y) P(y)  /  P(x)  &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; <br/><br/>
    Now, notice that the denominator P(x) doesn't depend on y, we can just get rid of it and write: <br/><br/>
    &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;  <strong>f<sub>opt</sub>(x) = argmax<sub>y</sub>
    P(x|y) P(y) &nbsp;&nbsp; &nbsp;&nbsp; (E2) </strong>  &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; 
    <strong style="color:rgb(0,120,191);">where f<sub>opt</sub> is the Bayes classifier</strong>
</p>
<p>
    <strong>Remember that we don't know any of these distributions so we're going to approximate them!</strong>
</p>

<h1 style="color:rgb(0,120,191);">3) Example with Gaussian class conditional densities</h1>
<p>
    Let's consider the following example: <br/>
    <ul>
        We get to observe scalar data (x &isin; &#8476;) and the corresponding label is either 0 or 1  (y &isin; {0, 1}). So it'a binary classification problem! <br/>
        Let's assume that <strong>we know the exact distribution of the (x, y) pairs</strong> in our dataset and that it's defined as follows: 
        <br/><br/>
        <li>The class prior for label 0 is &pi;<sub>0</sub> and the class prior for label 1 is &pi;<sub>1</sub></li>
        <li>The class conditionnal densities for classes 0 and 1 are univariate gaussians p<sub>0</sub>(x) &#8765; N(µ<sub>0</sub>, &sigma;<sub>0</sub><sup>2</sup>) &nbsp;&nbsp; and  &nbsp;&nbsp; p<sub>1</sub>(x) &#8765; N(µ<sub>1</sub>, &sigma;<sub>1</sub><sup>2</sup>) respectively.
    </ul>
</p>
<p>
    Let's recall the expression of the probability density of a univariate gaussian &#8765; f(x|µ, &sigma;<sup>2</sup>) 
    <img src="univariateGaussian.JPG" alt="P(x<sub>j</sub>|y) = (1/&sigma;&radic;(2&pi;))exp[-(x-µ)<sup>2</sup>/(2&sigma;<sup>2</sup>)]">
</p>
<p> The Bayes classifier assigns to each sample, the most probable class:
    <strong>f<sub>opt</sub>(x) = argmax<sub>y</sub>
    P(x|y) x P(y) &nbsp;&nbsp; &nbsp;&nbsp; (E2) </strong> <br/><br/>
    So for each data x, if p<sub>1</sub>(x)  &pi;<sub>1</sub> > p<sub>0</sub>(x) &pi;<sub>0</sub>, &nbsp; &nbsp; then class 1 will be assigned, otherwise, class 0 will be assigned!
    <br/><br/>
   Let's draw some graph illustrating this! <br/><br/>
</p>

In [1]:
import time
import numpy as np
import pandas as pd
import plotly.plotly as py
import plotly.graph_objs as go
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot

init_notebook_mode(connected=True)

x = np.linspace(-5,5,200)
pi0 = 1             # prior for class 0
µ0 = 0; sig0 = 1.5  # mean and variance of the gaussian class conditional density for class 0
pi1 = 0.5           # prior for class 1
µ1 = 1; sig1 = 0.5  # mean and variance of the gaussian class conditional density for class 1

prod0 = (1/(sig0*np.sqrt(2*np.pi))) * np.exp(-(x-µ0)**2/(2*sig0**2)) * pi0  # prod0 = likelihood0 * prior0
prod1 = (1/(sig1*np.sqrt(2*np.pi))) * np.exp(-(x-µ1)**2/(2*sig1**2)) * pi1  # prod1 = likelihood1 * prior1

line0 = go.Scatter(
    x = x,
    y = prod0,
    name='class 0'
)
line1 = go.Scatter(
    x = x,
    y = prod1,
    name='class 1'
)
anno0 = go.Scatter(
    x = [0.53, 0.53],
    y = [0,     0.45],
    mode = 'lines',
    textposition='bottom center',
    text = ['0.53'],
    name = 'boundary 1',
    line = dict(color = ('rgb(0, 0, 0)'),
        dash = 'dot')
)
anno1 = go.Scatter(
    x = [1.73, 1.73],
    y = [0,     0.45],
    mode = 'lines',
    text = ['1.73'],
    name = 'boundary 2',
    textposition='bottom center',
        line = dict(color = ('rgb(0, 0, 0)'),
        dash = 'dot')
)
layout = go.Layout(title='Likelihood times the prior, for both classes', showlegend=True)
data = [line0, line1, anno0, anno1]
fig = go.Figure(data=data, layout=layout)
iplot(fig)


<p>
    From the chart above, we see that class 1 is the most probable in the interval [0.53, 1.73] ! <br/>
    <strong>So for any data x &isin; [0.53, 1.73], class 1 will be predicted by the Bayes classifier! Outside this interval, the predicted class will be 0!</strong>
</p>

<h1 style="color:rgb(0,120,191);">4) The Plug In Classifier</h1>
<p>
    Once again, let's recall that we don't know the true distributions: that of the class conditionnal density and that of the prior are  unknown. <strong>So, in practice we use what we call a plug-in classifier: it consists of using the available data to approximate these unknown distributions, and therefore we will no longer be able to say that our classifier is optimal.</strong> (Remember that Bayes Classifier minimizes the prediction error) 
</p>
<br/>
<p>
    To illustrate this, consider a multiclass classification problem with K classes (K &ge;3), we observe n vectors x<sub>i</sub> (i = 1, ...,n) &isin;  &#8476;<sup>d</sup>. <br/>
    We will use Maximum Likelihood Estimation (MLE) to estimate the Bayes Classifier  :<br/>
<ul>
 <li><strong>The prior probability &pi;<sub>y</sub></strong>  that our observation is going to be labelled class y <strong>is the fraction of times that happens in the dataset.</strong> So we just need to sum up the number of times we see class y and divide by the number of observations:<br/>&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;
    <strong>&pi;<sub>y</sub> = (1/n) &sum;<sub>i&isin;[|1,n|]</sub>  1(y<sub>i</sub>=y)  &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;(E3)</strong>    
 </li><br/>
 <li><strong>The class conditionnal densities will also be estimated from the empirical dataset</strong> <br>
     Let's consider multivariate Gaussians for each class conditional density, P(x|y) &#8765; N(µ<sub>y</sub>, &Sigma;<sub>y</sub>). <br/>So we need to estimate the mean and the covariance. As for the prior, we will use the MLE: for each class y, we calculate the empirical mean of the data belonging to class y in the dataset, and the empirical covariance of the data belonging to class y in the dataset:<br/><br/>&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;
     <strong>µ<sub>y</sub> = (1/n<sub>y</sub>) <strong>&sum;</strong><sub>i&isin;[|1,n|]</sub>  1(y<sub>i</sub>=y) x<sub>i</sub>  &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; (E4)</strong><br/><br/>&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;
     
   <strong>&Sigma;<sub>y</sub> = (1/n<sub>y</sub>) <strong>&sum;</strong><sub>i&isin;[|1,n|]</sub>  1(y<sub>i</sub>=y) (x<sub>i</sub>-µ<sub>y</sub>) (x<sub>i</sub>-µ<sub>y</sub>)<sup>T</sup>&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; &nbsp; (E5)</strong>  <br/><br/>
     
 </li>
 
 <li> As seen before, for any new incoming data x, we take the product of the likelihood and the prior and choose the most probable class, thus the following expression of the Plug-In classifier (under the multivariate gaussian assumption for the class conditionnal densities): <br/><br/> &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;
    <strong>f(x) = argmax<sub>y&isin;&upsih;</sub>  &pi;<sub>y</sub> |&Sigma;<sub>y</sub>|<sup>-1/2</sup> exp{ (-1/2) (x-µ<sub>y</sub>)<sup>T</sup> &Sigma;<sub>y</sub><sup>-1</sup> (x-µ<sub>y</sub>) } &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; (E6)</strong> 
 </li>
 <p>
    Let's recall the expression of the probability density of a multivariate gaussian &#8765; f(x|µ, &Sigma;) 
    <img src="multivariateGaussian.JPG" alt="P(x<sub>j</sub>|y) = (1/&sigma;&radic;(2&pi;))exp[-(x-µ)<sup>2</sup>/(2&sigma;<sup>2</sup>)]">
</p>
</ul>
</p>


<h1 style="color:rgb(0,120,191);">5) Example using the Plug In Classifier</h1>
<p>
    We would like to determine wheter a given person is a male or a female, based on three features:<br/>
<ul>
    <li>The height (in feet)</li>
    <li>The weight (in lbs)</li>
    <li>The foot size (in inches)</li>
</ul>
Let's consider the following dataset: <br/>
</p>

<table>

<tbody><tr>
<th>Person</th>
<th>height (feet)</th>
<th>weight (lbs)</th>
<th>foot size(inches)
</th></tr>
<tr>
<td>male</td>
<td>6</td>
<td>180</td>
<td>12
</td></tr>
<tr>
<td>male</td>
<td>5.92</td>
<td>190</td>
<td>11
</td></tr>
<tr>
<td>male</td>
<td>5.58</td>
<td>170</td>
<td>12
</td></tr>
<tr>
<td>male</td>
<td>5.92</td>
<td>165</td>
<td>10
</td></tr>
<tr>
<td>female</td>
<td>5</td>
<td>100</td>
<td>6
</td></tr>
<tr>
<td>female</td>
<td>5.5</td>
<td>150</td>
<td>8
</td></tr>
<tr>
<td>female</td>
<td>5.42</td>
<td>130</td>
<td>7
</td></tr>
<tr>
<td>female</td>
<td>5.75</td>
<td>150</td>
<td>9
</td>
</tr>
    
</table>
<br/>

<p>
    Let's apply the formulas given previously to <strong>classify a new sample data of a person 5.8 feet high, 150lbs heavy and whose foot size is 10!</strong>
</p>
<br/>
<p>
    We have 2 classes "male" and "female", this a binary classification problem! <br/><br/>
    <strong>First, let's calculate the classes prior probabilities:</strong> <br/>
    In the dataset which comprises 8 observations, 4 are labelled male and 4 are labelled female
    so we have <br/>&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; <strong>(E3) &rArr; &nbsp;&nbsp;&pi;<sub>Male</sub> = &pi;<sub>Female</sub> = 4/8 = 0.5 </strong>
</p>    



<p>
    Under the multivariate Gaussian assumption for the likelihood, <strong>we now have to calculate the mean of each of the 2 classes </strong> "Male" and "Female",  <br/>&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;

<strong>(E4) &rArr; &nbsp;&nbsp;µ<sub>Male</sub> = (1/4) (x<sub>1</sub> + x<sub>2</sub> + x<sub>3</sub> + x<sub>4</sub>) </strong><br/><br/></p> 

In [2]:
piMale = 4/8
x1 = np.array([ [6], [180], [12] ])
x2 = np.array([ [5.92], [190], [11] ])
x3 = np.array([ [5.58], [170], [12] ])
x4 = np.array([ [5.92], [165], [10] ])
µMale = (x1+x2+x3+x4)/4
print( 'µMale = \n{}'.format(µMale) )


µMale = 
[[  5.855]
 [176.25 ]
 [ 11.25 ]]


&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;<strong>(E4) &rArr; &nbsp;&nbsp;µ<sub>Female</sub> = (1/4) (x<sub>5</sub> + x<sub>6</sub> + x<sub>7</sub> + x<sub>8</sub>) <br/><br/>

In [3]:
piFemale = 4/8
x5 = np.array([ [5], [100], [6] ])
x6 = np.array([ [5.5], [150], [8] ])
x7 = np.asarray([ [5.42], [130], [7] ])
x8 = np.asarray([ [5.75], [150], [9] ])
µFemale = (x5+x6+x7+x8)/4
print( 'µFemale = \n{}'.format(µFemale) )

µFemale = 
[[  5.4175]
 [132.5   ]
 [  7.5   ]]


<p>
Under the multivariate Gaussian assumption for the likelihood, <strong>we now have to calculate the variance of each of the 2 classes</strong> "Male" and "Female",  <br/>&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;
   <strong>(E5) &rArr; &nbsp;&nbsp;&Sigma;<sub>Male</sub> = (1/4) [ (x<sub>1</sub>-µ<sub>Male</sub>)(x<sub>1</sub>-µ<sub>Male</sub>)<sup>T</sup> +
    (x<sub>2</sub>-µ<sub>Male</sub>)(x<sub>2</sub>-µ<sub>Male</sub>)<sup>T</sup> +
    (x<sub>3</sub>-µ<sub>Male</sub>)(x<sub>3</sub>-µ<sub>Male</sub>)<sup>T</sup> +
    (x<sub>4</sub>-µ<sub>Male</sub>)(x<sub>4</sub>-µ<sub>Male</sub>)<sup>T</sup> ] </strong><br/>
</p>
<p> Let's rewrite this in matrix form to ease the notation:<br/>
&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;    <strong>&Sigma;<sub>Male</sub> = (1/4) [x<sub>1</sub> x<sub>2</sub> x<sub>3</sub> x<sub>4</sub>) - µ<sub>Male</sub>]   [ (x<sub>1</sub> x<sub>2</sub> x<sub>3</sub> x<sub>4</sub>) - µ<sub>Male</sub> ]<sup>T</sup></strong>

In [4]:
SigmaMale = (1/4) * np.dot( np.hstack((x1,x2,x3,x4)) - µMale, (np.hstack((x1,x2,x3,x4)) - µMale).T ) 
print( 'SigmaMale = \n{}'.format(SigmaMale) )

SigmaMale = 
[[ 2.62750e-02  6.06250e-01 -4.87500e-02]
 [ 6.06250e-01  9.21875e+01  2.18750e+00]
 [-4.87500e-02  2.18750e+00  6.87500e-01]]


<p>
&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;    <strong>(E5) &rArr; &Sigma;<sub>Female</sub> = (1/4) [ (x<sub>5</sub>-µ<sub>Female</sub>)(x<sub>5</sub>-µ<sub>Female</sub>)<sup>T</sup> +
    (x<sub>6</sub>-µ<sub>Female</sub>)(x<sub>6</sub>-µ<sub>Female</sub>)<sup>T</sup> +
    (x<sub>7</sub>-µ<sub>Female</sub>)(x<sub>7</sub>-µ<sub>Female</sub>)<sup>T</sup> +
    (x<sub>8</sub>-µ<sub>Female</sub>)(x<sub>8</sub>-µ<sub>Female</sub>)<sup>T</sup> ] </strong><br/>
</p>
<p> Let's rewrite this in matrix form to ease the notation:<br/>
&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;    <strong>&Sigma;<sub>Female</sub> = (1/4) [x<sub>5</sub> x<sub>6</sub> x<sub>7</sub> x<sub>8</sub>) - µ<sub>Female</sub>]   [ (x<sub>5</sub> x<sub>6</sub> x<sub>7</sub> x<sub>8</sub>) - µ<sub>Female</sub> ]<sup>T</sup></strong>

In [5]:
SigmaFemale = (1/4) * np.dot( np.hstack((x5,x6,x7,x8)) - µFemale, (np.hstack((x5,x6,x7,x8)) - µFemale).T ) 
print( 'SigmaFemale = \n{}'.format(SigmaFemale) )

SigmaFemale = 
[[7.291875e-02 5.206250e+00 2.912500e-01]
 [5.206250e+00 4.187500e+02 2.125000e+01]
 [2.912500e-01 2.125000e+01 1.250000e+00]]


<p>
    We can now calculate the product of the likelihood and the prior for both classes Male and Female, evaluated for the vector x = (6 130 8)<sup>T</sup> :<br/>
</p>      
<ul> 
<li>      &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;    <strong>
    &pi;<sub>Male</sub> |&Sigma;<sub>Male</sub>|<sup>-1/2</sup> exp{ (-1/2) (x-µ<sub>Male</sub>)<sup>T</sup> &Sigma;<sub>Male</sub><sup>-1</sup> (x-µ<sub>Male</sub>) } </strong> </li> 
<li>      &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;    <strong>
    &pi;<sub>Female</sub> |&Sigma;<sub>Female</sub>|<sup>-1/2</sup> exp{ (-1/2) (x-µ<sub>Female</sub>)<sup>T</sup> &Sigma;<sub>Female</sub><sup>-1</sup> (x-µ<sub>Female</sub>) } </strong> </li>  
<ul> 
  

In [6]:
xt = np.array([ [5.8], [150], [10] ])

def calcPosteriorNum(x, pi, µ, Sigma):
    factor = np.linalg.det(Sigma)**(-1/2)
    invSigma = np.linalg.inv(Sigma)
    expTerm = (-1/2) * np.dot( np.dot((x - µ).T, invSigma), x - µ )
    posteriorNum = factor * np.exp(expTerm) * pi
    return posteriorNum[0][0]

posteriorNumMale = calcPosteriorNum(xt, piMale, µMale, SigmaMale)
print('Numerator of the posterior probability for x being a Male = {}'.format(posteriorNumMale))

posteriorNumFemale = calcPosteriorNum(xt, piFemale, µFemale, SigmaFemale)
print('Numerator of the posterior probability for x being a Female = {}\n'.format(posteriorNumFemale))

if posteriorNumMale > posteriorNumFemale:
    print('posteriorNumMale > posteriorNumFemale, So the predicted class is Male!\n')
else:
    print('posteriorNumMale < posteriorNumFemale, So the predicted class is Female!\n')
        


Numerator of the posterior probability for x being a Male = 0.008199794869542944
Numerator of the posterior probability for x being a Female = 0.00012255000541263892

posteriorNumMale > posteriorNumFemale, So the predicted class is Male!



<p>
<strong>So according to our calculations, the class predicted for our new value x = (5.8 130 8)<sup>T</sup> is Female ! </strong>
</p>

<h1 style="color:rgb(0,120,191);">6) A more generalist implementation of the Bayes Plug-In Classifier</h1>

In [7]:
def pluginClassifier(X_train, y_train, X_test):    
    """
    X_train is a n by d matrix of observations, n is the number of observations, d is the dimension
    y_train is a n by 1 numeric vector of the labels
    X_test is a m by d matrix of the data to classify, m is the number of observations to classify, d is the dimension
    """
    # 1st, compute the class prior probability
    nbClasses = len(np.unique(y_train))
    allClasses = range(0,nbClasses)
    nTrainObs = X_train.shape[0]   # number of observations in the train set
    nTestObs = X_test.shape[0]     # number of observations in the test set
    priorProb = []        # probability of the prior 
    clX  = {}             # dictionnary containing the covariates belonging to a certain class (the key is the class)
    muX  = {}             # dictionnary containing the mean of the vectors labelled to a certain class (the key is the class)
    varX = {}             # dictionnary containing the variance of the vectors labelled to a certain class (the key is the class)
    numPostX = np.ones( (nTestObs, len(allClasses)) )  # numpy array containing the numerator of the posterior probablities
    probX = np.ones( (nTestObs, len(allClasses)) )     # numpy array containing the numerator of the posterior probablities    
    assignedClass = np.zeros((nTestObs,1),int)             # numpy array containing the assigned classes
    for cl in allClasses:
        ny = len( y_train[y_train == cl] )
        curClassPriorProb =  ny/nTrainObs
        priorProb.append(curClassPriorProb)
      
        # Retrieve the inputs from X_train which belong to class cl
        X_train_df = pd.DataFrame(X_train)
        clX[cl] = X_train_df[y_train == cl]
        # Retrieve the numpy array and transpose it to have a column vector
        clX[cl] = np.transpose(clX[cl].values)
        assert(ny == clX[cl].shape[1])
     
        # Compute mean of the inputs which belong to class cl
        
        muX[cl] = (1/ny) * np.sum( clX[cl], axis=1 )
        muX[cl] = (muX[cl]).reshape(X_train.shape[1],1)
        assert( muX[cl].shape == (X_train.shape[1],1) )


        # Compute variance of theses inputs which belong to class cl
        ecart = np.dot( (clX[cl] - muX[cl]), np.transpose(clX[cl] - muX[cl]) )
        #print(ecart)
        varX[cl] = (1/ny) * ecart
        assert(varX[cl].shape[0] == varX[cl].shape[1])
        assert(X_train.shape[1] == varX[cl].shape[0])
        deter = np.linalg.det(varX[cl])    # determinant of the variance matrix
        inversVar = np.linalg.inv(varX[cl])

        # Parse the test vector to compute the probability of the Bayes Plug-In classifier
        for idx in range(nTestObs) :
            curTestObs = np.transpose( [X_test[idx,:]] )  # make it a row vector
            expTerm = (-1/2) * np.dot( np.transpose(curTestObs-muX[cl]), np.dot(inversVar, curTestObs-muX[cl]) )
            factor = (1/(2*np.pi))**(X_train.shape[1]/2) * (1/np.sqrt(deter)) #
            numPostX[idx,cl] =  factor * np.exp(expTerm) * priorProb[cl]

    
    
    # Create Probabilities that sum to 1
    for idx in range(nTestObs) :
        rescale = 1/np.sum(numPostX[idx,:])
        probX[idx,:] = rescale * numPostX[idx,:]
        assignedClass[idx] = numPostX[idx,:].tolist().index(max(numPostX[idx,:]))
#     probXList = probX.tolist()
  
    return assignedClass, numPostX, probX
 



In [8]:
X_train = np.transpose(np.hstack((x1,x2,x3,x4,x5,x6,x7,x8)))
y_train = np.array([ [0], [0], [0], [0], [1], [1], [1], [1] ])   # 0 is the male class, 1 is the female class
X_test = np.transpose(xt)      # make it a row vector
assignedClass, numPostX, probX = pluginClassifier(X_train, y_train, X_test) # assuming final_outputs is returned from function

In [9]:
assignedClass[0][0]
if assignedClass[0][0] == 0:
    print('\n Class predicted by our plug-in classifier function is Male!\n')
else:
    print('\n Class predicted by our plug-in classifier function is Female!\n')


 Class predicted by our plug-in classifier function is Male!



<h1 style="color:rgb(0,120,191);">7) The Naïve Bayes Classifier</h1>
<p>
    We have presented the Bayes classifier which is optimal in the sense that it minimizes the prediction error, but requires the knowledge of the true distribution of the data.
</p> 
<p>
As in practice we don't know this distribution, we make an assumption about it and we use the available data to learn the parameters of the chosen distribution, this leads to the Plug-In classifier, which computation is pretty heavy and requires inversing a matrix.
</p>
<p>
    Now can we make a simplifying asumption to ease the calculations? <br/>
    The answer is yes, let's suppose that the distribution of the features given the class is independant. 
    This is called <strong style="color:rgb(0,120,191);">"Naïve"</strong> because we're breaking any correlations in the feature, which is a naïve assumption, but in the facts, <strong> it works fairly well!</strong>.
</p>
<br/>
<h3 style="color:rgb(0,120,191);"> Finding the new expresion of our classifier with the Naïve Bayes assumption</h3>
<p> For any vector x among our observations, we want to assign the most probable class. <br/><strong>This is done by taking the  argmax<sub>y</sub> P(x|y) P(y) &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;  (E2)</strong> <br/><br/>
    x is a vector of d features: x = [x<sub>1</sub> x<sub>2</sub> x<sub>3</sub> ...x<sub>d</sub> ] where <strong   style="color:rgb(0,120,191);"> each x<sub>j</sub> are independant features given the class</strong> (because of the Naïve Bayes assumption)<br/>
    <strong>Remember that the product P(x|y) P(y) is equal to the joint probability P(x,y)!</strong> <br/>
    <p>
&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;        P(x,y) = P(x<sub>1</sub>, x<sub>2</sub>, ..., x<sub>d</sub>, y) <br/>
&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;        P(x,y) = P(x<sub>1</sub>|x<sub>2</sub>, ..., x<sub>d</sub>, y)  P(x<sub>2</sub>, ..., x<sub>d</sub>, y) <br/>
&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;        P(x,y) = P(x<sub>1</sub>|x<sub>2</sub>, ..., x<sub>d</sub>, y)  P(x<sub>2</sub>|x<sub>3</sub>, ..., x<sub>d</sub>, y) P(x<sub>3</sub>, ..., x<sub>d</sub>, y) <br/>
&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;        ... <br/>
&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;        P(x,y) = P(x<sub>1</sub>|x<sub>2</sub>, ..., x<sub>d</sub>, y)  P(x<sub>2</sub>|x<sub>3</sub>, ..., x<sub>d</sub>, y) ... P(x<sub>d-1</sub>|x<sub>d</sub>, y) P(x<sub>d</sub>|y)P(y) <br/> <br/>
  Now the naïve conditional independance assumption comes into play: we assume that each feature x<sub>j</sub> is conditionally independant of every other feature x<sub>m</sub> <br/>for j &ne; m, given the category y. Then<br/> 
        &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;        P(x<sub>j</sub>|x<sub>j+1</sub>, x<sub>j+2</sub>, ..., x<sub>d</sub>, y) = P(x<sub>j</sub>|y)   &nbsp; &nbsp;&nbsp; And we can write <br />
        &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;        P(x,y) = P(x<sub>1</sub>|y)  P(x<sub>2</sub>|y) ... P(x<sub>d-1</sub>| y) P(x<sub>d</sub>|y) P(y) <br/> 
        &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;        P(x,y) = P(y) &#928;<sub>j &isin;{1,d}</sub> P(x<sub>j</sub>|y)
  </p>
    <p>
    Eventually, the Naïve Bayes classifier is the function that assings a class label as follows: <br/> &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp;    <strong>argmax<sub>y</sub> P(y) &#928;<sub>j &isin;{1,...,d}</sub> P(x<sub>j</sub>|y) &nbsp;&nbsp; (E7)</strong>
    
</p>

<h1 style="color:rgb(0,120,191);">8) Naïve Bayes Applied to the gender classification problem</h1>
<p>
    For our gender classification problem, we have 3 features (x<sub>j</sub>): height, weight and footsize. <br/>
    To classify a new sample data of a person 5.8 feet high, 150lbs heavy and whose foot size is 10, we derive from (E7) that we need to calculate the two products:
    <ul>
        <li>P(Male) P(height|Male) P(weight|Male) P(footsize|Male)</li>
        <li>P(Female) P(height|Female) P(weight|Female) P(footsize|Female)</li>
</ul>
And decide the class of the point as being the most probable!
</p>    

<p>
   Remember the dataset: <br/>
</p>

<table>

<tbody><tr>
<th>Person</th>
<th>height (feet)</th>
<th>weight (lbs)</th>
<th>foot size(inches)
</th></tr>
<tr>
<td>male</td>
<td>6</td>
<td>180</td>
<td>12
</td></tr>
<tr>
<td>male</td>
<td>5.92</td>
<td>190</td>
<td>11
</td></tr>
<tr>
<td>male</td>
<td>5.58</td>
<td>170</td>
<td>12
</td></tr>
<tr>
<td>male</td>
<td>5.92</td>
<td>165</td>
<td>10
</td></tr>
<tr>
<td>female</td>
<td>5</td>
<td>100</td>
<td>6
</td></tr>
<tr>
<td>female</td>
<td>5.5</td>
<td>150</td>
<td>8
</td></tr>
<tr>
<td>female</td>
<td>5.42</td>
<td>130</td>
<td>7
</td></tr>
<tr>
<td>female</td>
<td>5.75</td>
<td>150</td>
<td>9
</td>
</tr>
    
</table>
<br/>

<p>
    Let's apply the formulas given previously to classify a new sample data of a person 5.8 feet high, 150lbs heavy and whose foot size is 10!
</p>
<p>
    We have 2 classes "male" and "female", this a binary classification problem! <br/>
    First, let's calculate the classes prior probabilities: <br/>
    In the dataset which comprises 8 observations, 4 are labelled male and 4 are labelled female
    so we have <br/>&nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; <strong>&pi;<sub>Male</sub> = &pi;<sub>Female</sub> = 4/8 = 0.5      &nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; ( P(Male) = &pi;<sub>Male</sub>  and P(Female) = &pi;<sub>Female</sub> )</strong>
</p>    
<p>
    Then, to estimate the probability densities P(x<sub>j</sub>|y), we can make the assumption they follow univariate gaussian distribution P(x<sub>j</sub>|y) &#8765; f(x|µ, &sigma;<sup>2</sup>) 
    <img src="univariateGaussian.JPG" alt="P(x<sub>j</sub>|y) = (1/&sigma;&radic;(2&pi;))exp[-(x-µ)<sup>2</sup>/(2&sigma;<sup>2</sup>)]">
</p>

<p>
    So P(height|Male) &#8765; N(µ<sub>0</sub>, &sigma;<sub>0</sub><sup>2</sup>), P(weight|Male)  &#8765; N(µ<sub>1</sub>, &sigma;<sub>1</sub><sup>2</sup>), P(footsize|Male)  &#8765; N(µ<sub>2</sub>, &sigma;<sub>2</sub><sup>2</sup>) <br/><br/>
    From the dataset, we can derive µ0 = (6+5.92+5.58+5.92)/4 = 5.855 &nbsp;&nbsp;&nbsp;&nbsp; and &nbsp;&nbsp;&nbsp;&nbsp; &sigma;<sub>0</sub><sup>2</sup> = var(6+5.92+5.58+5.92) &asymp; 0.02675  <br/>
    Injecting these values into the equation above leads to P(height|Male)  &asymp; 2.32 <br/>
    
   <br/> To make our calculations easier, let's define a function to compute the output of the univariate gaussian distribution evaluated in one point
</p>    

In [10]:
def uniGauss(µ, var, x):
    pdfx = (1/np.sqrt(2*np.pi*var)) * np.exp( -(x-µ)**2/(2*var) )
    return pdfx


<p>
<strong>Let's proceed the same way to get the other probabilities:</strong>
</p>

In [11]:
# Calculating the posterior numerator for the Male class
l0 = [6, 5.92, 5.58, 5.92]
µ0 = np.mean(l0)
var0 = np.std(l0)**2
pHeightGivenMale = uniGauss(µ0, var0, xt[0][0])

l1 = [180, 190, 170, 165]
µ1 = np.mean(l1)
var1 = np.std(l1)**2
pWeightGivenMale = uniGauss(µ1, var1, xt[1][0])

l2 = [12, 11, 12, 10]
µ2 = np.mean(l2)
var2 = np.std(l2)**2
pFootSizeGivenMale = uniGauss(µ2, var2, xt[2][0])
posteriorNumMale = piMale * pHeightGivenMale * pWeightGivenMale * pFootSizeGivenMale


# Calculating the posterior numerator for the Female class
l3 = [5, 5.5, 5.42, 5.75]
µ3 = np.mean(l3)
var3 = np.std(l3)**2
pHeightGivenFemale = uniGauss(µ3, var3, xt[0][0])

l4 = [100, 150, 130, 150]
µ4 = np.mean(l4)
var4 = np.std(l4)**2
pWeightGivenFemale = uniGauss(µ4, var4, xt[1][0])

l5 = [6, 8, 7, 9]
µ5 = np.mean(l5)
var5 = np.std(l5)**2
pFootSizeGivenFemale = uniGauss(µ5, var5, xt[2][0])
posteriorNumFemale = piFemale * pHeightGivenFemale * pWeightGivenFemale * pFootSizeGivenFemale

print('posteriorNumMale = {}'.format(posteriorNumMale))
print('posteriorNumFemale = {}\n'.format(posteriorNumFemale))

if posteriorNumMale > posteriorNumFemale:
    print('posteriorNumMale > posteriorNumFemale, So our Naïve Bayes Classifier labels our new sample as belonging to the Male class!\n')
else:
    print('posteriorNumMale < posteriorNumFemale, So our Naïve Bayes Classifier labels our new sample as belonging to the Female class!\n')

posteriorNumMale = 0.00017756467590579106
posteriorNumFemale = 0.00010730313438090195

posteriorNumMale > posteriorNumFemale, So our Naïve Bayes Classifier labels our new sample as belonging to the Male class!



<h1 style="color:rgb(0,120,191);">9) Plug-In Bayes Classifier Applied to the iris dataset</h1>
<p>
    Let's apply our implementation of the Plug-In classifier on the well known "iris dataset". <br/> 
 The data set contains 3 classes of 50 instances each, where each class refers to a type of iris plant.<br/>
    The predicted attribute is the class of iris plant. <br/><br/>
<strong>Attribute Information:</strong>
<ol>       <li>   sepal length in cm </li>
<li>   sepal width in cm </li>
<li>   petal length in cm </li>
<li>   petal width in cm </li>
<li>   class:  Iris Setosa, Iris Versicolour, Iris Virginica</li>    
</ol>    

In [12]:
# First read the data
data = pd.read_csv('iris.data', header=None, names=['sepalLen', 'sepalWi', 'petalLen', 'petalWi','label'], index_col=False )
data.dtypes

sepalLen    float64
sepalWi     float64
petalLen    float64
petalWi     float64
label        object
dtype: object

In [13]:
# Create a function to assign a numeric value to the labels
def createNumericClass(inputString):
    # for each string assign a numeric label
    if inputString == 'Iris-setosa':
        val = 0
    elif inputString == 'Iris-versicolor':
        val = 1
    elif inputString == 'Iris-virginica':
        val = 2
    else:
        val = -1
    return val

data['target'] = data['label'].apply(lambda x: createNumericClass(x))

data.columns

Index(['sepalLen', 'sepalWi', 'petalLen', 'petalWi', 'label', 'target'], dtype='object')

In [14]:
# Train the classifier and predict classes
from sklearn.metrics import accuracy_score

dataSamp = data.sample(frac=1, random_state=1, axis=0)        # shuffle the dataset 
propTrain = 0.45                                              # Proportion of data to keep for the training
nbSampleTrain = int(propTrain*dataSamp.shape[0])              # Equivalent number of data
XTrain = dataSamp.iloc[:nbSampleTrain,:4].values              # Data for the training set (without the labels)
yTrain = dataSamp.iloc[:nbSampleTrain,5].values               # Labels for the training set
XVal = dataSamp.iloc[nbSampleTrain:,:4].values                # Data for the validation set
yVal = dataSamp.iloc[nbSampleTrain:,5].values                 # Labels for the validation set

assignedClass, numPostX, probX = pluginClassifier(XTrain, yTrain, XVal)

acc = accuracy_score(yVal, assignedClass)
print('Accuracy on test set = {:.2f}'.format(acc))
A = np.concatenate((yVal.reshape(yVal.shape[0],1), assignedClass), axis=1)



Accuracy on test set = 0.95


In [15]:
# Plot the Classification results on the validation set

trueClass = go.Scatter(
    x = dataSamp.index[nbSampleTrain:],
    y = yVal,
    mode = 'markers',
    name = 'True Class')

plugInPred = go.Scatter(
    x = dataSamp.index[nbSampleTrain:],
    y = assignedClass.T[0],
    mode = 'markers',
    name = 'Plug-In Classifier')

layout = go.Layout(title='Iris dataset Validation, Accuracy = {:.1f}%'.format(100*acc), showlegend=True)
data = [trueClass, plugInPred]
fig = go.Figure(data=data, layout=layout)
iplot(fig)

<h1 style="color:rgb(0,120,191);">10) Naïve Bayes Applied to the iris dataset</h1>

In [16]:
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB()                                         # Create an instance of the Naïve Bayes classifier
clf.fit(XTrain, yTrain)                                    # Train the model
NBPred = clf.predict(XVal)                                   # Make predictions
accNB = accuracy_score(yVal, NBPred)                         # Compute the accuracy score
print('Accuracy on test set = {:.2f}'.format(accNB))       # Print the score

Accuracy on test set = 0.95


In [17]:
# Plot the Classification results on the validation set

NaiveBayesPred = go.Scatter(
    x = dataSamp.index[nbSampleTrain:],
    y = NBPred,
    mode = 'markers',
    name = 'Naïve Bayes Classifier')

layout = go.Layout(title='Iris dataset Validation, Accuracy = {:.1f}%'.format(100*acc), showlegend=True)
data = [trueClass, plugInPred, NaiveBayesPred]
fig = go.Figure(data=data, layout=layout)
iplot(fig)

<h1 style="color:rgb(0,120,191);">11) Conclusion</h1>
<p>
    Through this tutorial, we explored the Bayes classifier, the plug-in classifier and the Naïve Bayes classifier. <br/>
  <ul>
     <li>    We gave proper definition of each of these classifiers and played with a toy example regarding gender classification.</li>
     <li>We also derived our own implementation of the plug-in classifier</li>
     <li>We enventually achieved a fairly reasonnable accuracy of 95% during the validation step on the iris dataset</li>
    </ul>   
</p>

<p>
    Following ressources were used to build this tutorial:
    <ul>
        <li><a href="https://web.stanford.edu/~hastie/Papers/ESLII.pdf">The elements of statistical learning</a></li>
        <li><a href="https://www.edx.org/course/machine-learning">Columbia Machine Learning Course</a></li>
        <li><a href="https://en.wikipedia.org/wiki/Naive_Bayes_classifier">Gender classification exemple</a></li>
        <li><a href="https://archive.ics.uci.edu/ml/datasets/Iris/">UCI Machine Learning Repository</a></li>
    </ul>
</p>
<br/>
<p>
Author: Chamsdine SADIKOU <br/>
Company: CS CONTROLCOM    <br/>
Version: 1.0.0    
</p>